Added missing status code 429 Too Many Requests

KeithHenry opened the pull request:
This enumeration is missing HTTP status code 429 Too Many Requests - this is used by many rate-limited services to indicate that the current allocation has been exceeded.

For example the British government's companies registry service does this: https://developer.companieshouse.gov.uk/api/docs/index/gettingStarted/rateLimiting.html

.NET should be able to handle and create responses with this status code.

KeithHenry committed:
Added missing status code 429 Too Many Requests

This enumeration is missing HTTP status code 429 Too Many Requests - this is used my many rate-limited services to indicate that the current allocation has been exceeded.

For example the British governments companies registry service does this: https://developer.companieshouse.gov.uk/api/docs/index/gettingStarted/rateLimiting.html

.NET should be able to handle and create responses with this status code.

JonHanna commented on src/System.Net.Primitives/src/System/Net/HttpStatusCode.cs:
```diff
@@ -56,6 +56,7 @@ public enum HttpStatusCode
         ExpectationFailed = 417,
 
         UpgradeRequired = 426,
+        TooManyRequests = 429,
```
If adding one status from RFC 6585, it seems reasonable to add the rest (428, 431 & 511) at the same time.

KeithHenry replied:
It does, though only 429 is causing me a problem. Shall I add that commit?

KeithHenry replied:
@JonHanna done, cheers for the suggestion :-)

KeithHenry committed:
Added all the additional codes from RFC 6585

Following comment on PR added all the codes from https://tools.ietf.org/html/rfc6585

davidsh commented:
cc: @karelz 

Thx for the PR.  However, this is an API change and would require changes to the REF folder.  In addition, there is an API review process that needs to be followed first.

We are also tracking adding new values to this enumeration here: #4382.

So, it probably be best to close that PR and to use issue #4382 to propose the full set of additions to this enumeration.

KeithHenry committed:
Added new HTTP status codes to REF

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=26415) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=26415) before we can accept your contribution.<br/><br/>:x: KeithHenry [sign now](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=26415)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/corefx?pullRequest=26415) it.</sub>

KeithHenry commented:
Cheers @davidsh, I've updated the copy in the ref folder. 

Adding enums is an API change? If so it has to be one that can change quickly (well, relatively, [RFC 6585](https://tools.ietf.org/html/rfc6585) being from 2012 and #4382 from 2015), and while HTTP codes are a standard has it been decided that .NET's `HttpClient` can only consume services that absolutely stick to that standard?

To be honest I think this _needs_ an API change. I mean that `HttpClient` _should_ be able to handle _any_ HTTP status code, even arbitrary ones made up by wilfully bad APIs, and an `enum` be something that helps develop for standard ones. 

I just want to get this bug fixed (and it is a bug).

karelz commented:
@KeithHenry Touching any public surface is API change. API review checks also name consistency with other BCL types, etc.

> If so it has to be one that can change quickly (well, relatively, RFC 6585 being from 2012 and #4382 from 2015), and while HTTP codes are a standard has it been decided that .NET's HttpClient can only consume services that absolutely stick to that standard?
> To be honest I think this needs an API change. I mean that HttpClient should be able to handle any HTTP status code, even arbitrary ones made up by wilfully bad APIs, and an enum be something that helps develop for standard ones.

`HttpClient` is built on top of system libraries - WinHTTP and libcurl. They do most of the handling and as such define most of the "standard". This enum does not limit functionality of `HttpClient`, it is a convenience.

> I just want to get this bug fixed (and it is a bug).

Definition of bug vs. feature is often biased opinion. Nevertheless, we should run it by API review first. It may take a week or two as we are super-busy reviewing APIs critical for 2.1 at this moment (just to set expectations).

KeithHenry commented:
@karelz am I missing something?

> This enum does not limit functionality of HttpClient, it is a convenience.

This started because I have an external API that returns code 429 once a limit has been reached - either WinHTTP or cURL could handle that result, but `HttpClient` can't - there is no value that `HttpResponseMessage.StatusCode` can hold.

Is there some way to get the underlying code from `HttpClient` that I'm missing?

> Touching any public surface is API change

I understand that, but this enum is not coming from a source internal to .NET - is this .NET's API to regulate?

> It may take a week or two

#4382 is 3 years old. Is there a way to push for a review sooner or help it along?

davidsh commented:
@KeithHenry Thank you very much for your PR on this.  The .NET Core team was able to close on the expanded set of HttpStatusCode values to add to the enum as part of a final API review #4382

A new PR #26727  was opened up that now supersedes this PR.  So, I'll close this PR.

lindexi commented:
See https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Primitives/src/System/Net/HttpStatusCode.cs
